### ISSUE TITLE
structs/enums questions for new PG 13 upgrade

### ISSUE THREAD

**Issue by pyramation** [pganalyze/libpg_query#89]
@lfittl so happy to see the PG 13 launch! congrats 🎉

I'm starting to rebuild the node deparser and leveraging the enums and structs json files. I'm hoping to try and automate some parts by leveraging these files and wanted to make sure my assumptions are correct.

1) I noticed sometimes there are (pointers?) to objects and other times the objects themselves. Are they potentially arrays of objects when there is a "*" or can I just treat either `Thing` and `Thing*` as a single nested `Node` (without the wrapper)?

2) `Node` and `List` I imagine are the typical `Node` object which we can parse as before, and then `List` is an array of `Nodes`

Here is an example I pulled out demonstrating `Expr` and `Expr*` is used:

``` 

        {
          name: 'xpr',
          c_type: 'Expr',
          comment: null
        },

...

        {
          name: 'aggdistinct',
          c_type: 'List*',
          comment: '/* DISTINCT (list of SortGroupClause) */'
        },
        {
          name: 'aggfilter',
          c_type: 'Expr*',
          comment: '/* FILTER expression, if any */'
        },
        {
          name: 'aggstar',
          c_type: 'bool',
          comment: "/* true if argument list was really '*' */"
        },

...

        {
          name: 'exprnulls',
          c_type: 'bool*',
          comment: null
        },


```

**Comment by lfittl** [comment]
> @lfittl so happy to see the PG 13 launch! congrats 🎉
> 
> I'm starting to rebuild the node deparser and leveraging the enums and structs json files. I'm hoping to try and automate some parts by leveraging these files and wanted to make sure my assumptions are correct.

Excellent!

> 1. I noticed sometimes there are (pointers?) to objects and other times the objects themselves. Are they potentially arrays of objects when there is a "*" or can I just treat either `Thing` and `Thing*` as a single nested `Node` (without the wrapper)?

You can just treat these as singular objects - its mostly an implementation detail at the C level, from an output perspective these are handled the same.

> 2. `Node` and `List` I imagine are the typical `Node` object which we can parse as before, and then `List` is an array of `Nodes`

Thats correct - List nodes are explicitly called out as such, and always contain a list of Nodes. Note there are special cases like IntList and OidList, but I think they don't really surface in raw parse trees as seen in pg_query.

You can also look at the Protobuf definition for comparison, which shows how it gets simplified there (automatically created based on the same source file):

https://github.com/pganalyze/libpg_query/blob/6517eedf6c3c6c53a14ecd8f01410bb8fc3c8ec1/protobuf/pg_query.proto#L365-L367

**Comment by pyramation** [comment]
Great! That's super clear. Thank you!

**Comment by pyramation** [comment]
Seems that I made it through, time to sleep :) I ended up writing a small preparser to convert List and items back into arrays so the deparser could behave the same, but did upgrade to all the latest structs and enums peppered throughout the deparser. The new enums are going to be GREAT for future upgrades, that was a great idea.

One other small thing I noticed was that empty nodes were `null` now are `{}`, but my tests caught that change. 

Thanks again!
